File: PHP/calculator-tests.php
<?php /*--------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

namespace Liersch\Calculator;

require_once dirname(__FILE__) . '/calculator.php';

use \InvalidArgumentException;

function performTests(array &$tests): bool
{
  $testCount = 0;
  $errorCount = 0;
  $calculator = new Calculator(new FloatFormatter(DEFAULT_FLOAT_FORMAT_16));
  foreach ($tests as $test)
  {
    echo 'Test: ' . $test->name . "\n";
    $array = is_array($test->expression) ? $test->expression : array($test->expression);
    foreach ($array as $expr)
    {
      $lines = $calculator->calculateAndFormat($expr, '');
      if (!assertEqual($test->output, $lines))
        $errorCount++;
      $testCount++;
    }
  }
  echo "=> $testCount tests completed with $errorCount errors\n";
  return $errorCount <= 0;
}

function assertEqual(array &$expected, array &$actual): bool
{
  if (!is_array($expected))
    throw new InvalidArgumentException('Array expected (1)');

  if (!is_array($actual))
    throw new InvalidArgumentException('Array expected (2)');

  $c1 = count($expected);
  $c2 = count($actual);
  if ($c1 != $c2)
  {
    echo "Assertion failed: Arrays with different lengths\n";
    return false;
  }

  for ($i = 0; $i < $c1; $i++)
  {
    if ($expected[$i] != $actual[$i])
    {
      echo "Assertion failed: {$expected[$i]} != {$actual[$i]})\n";
      return false;
    }
  }

  return true;
}

File: Python/calculator_tests.py
#:----------------------------------------------------------------------------
#:
#:  Copyright © 2023 Steffen Liersch
#:  https://www.steffen-liersch.de/
#:
#:----------------------------------------------------------------------------

from typing import Any, List
from calculator import Calculator
from float_formatter import FloatFormatter


def perform_tests(tests: Any) -> bool:
  test_count = 0
  error_count = 0
  calculator = Calculator(FloatFormatter('{:.16G}'))
  for test in tests:
    print('Test: ' + test['name'])
    array = test['expression'] if type(test['expression']) == list else [test['expression']]
    for expr in array:
      lines = calculator.calculate_and_format(expr, '')
      if not assert_equal(test['output'], lines):
        error_count += 1
      test_count += 1
  print(f'=> {test_count} tests completed with {error_count} errors')
  return error_count <= 0


def assert_equal(expected: List[str], actual: List[str]) -> bool:
  if type(expected) != list: raise ValueError('Array expected (1)')
  if type(actual) != list: raise ValueError('Array expected (2)')

  c1 = len(expected)
  c2 = len(actual)
  if c1 != c2:
    print('Assertion failed: Arrays with different lengths', expected, actual)
    return False

  ok = all([a == b for a, b in zip(actual, expected)])
  if not ok:
    print('Assertion failed')
    return False

  return True
